### ISSUE TITLE
WAL segments before first backup on new S3 bucket immediately expired

### ISSUE THREAD

**Issue by acheccuc-tw** [pgbackrest/pgbackrest#1000]
Please provide the following information when submitting an issue (feature requests or general comments can skip this):

1. pgBackRest version: 2.24

2. PostgreSQL version: 9.6 (but applied to any version)

3. Operating system/version - if you have more than one server (for example, a database server, a repository host server, one or more standbys), please specify each: 
Ubuntu 18.04LTS on all servers

4. Did you install pgBackRest from source or from a package?
deb package

5. Please attach the following as applicable:
```
[db]
pg1-path=/pg_data/9.6/b
pg1-port=5432
pg1-host=db1
pg2-path=/pg_data/9.6/db
pg2-port=5432
pg2-host=db3
pg3-path=/pg_data/9.6/db
pg3-port=5432
pg3-host=db2
retention-full=208
retention-diff=416
retention-archive=14
process-max=6

[global]
process-max=6
repo1-s3-endpoint=xxx
repo1-s3-bucket=xxx
repo1-s3-key=xxx
repo1-s3-key-secret=xxx
repo1-s3-region=xxx
repo1-type=s3
stop-auto=y
backup-standby=y
start-fast=y
log-level-console=error
log-level-file=detail
log-path=/var/log/pgbackrest/
compress-level=3
archive-timeout=120
buffer-size=16777216
```

7. Describe the issue:
For compliance reasons we had to change the S3 bucket we used to take backups. We decided not to copy the old backups over (100 of TBs), but rather let them become obsolete (according to our recovery window) before dropping the old S3 bucket.
The process we used to switch the S3 buckets:

- change bucket configuration on backup server
- create stanza on the new S3 bucket
- change S3 bucket configuration on PostgreSQL replicas
- stop archiving on primary (archive_command to '/bin/false')
- change bucket configuration on primary
- re-enable archiving on primary
- run a check on backup server to confirm that archiving is performed
- trigger a full backup from backup server

After that we tried to perform a restore, starting from a backup on the old S3 bucket:

- start a restore on a new PostgreSQL server using a backup from the old bucket
- let PostgreSQL instruct pgbackrest to restore and recover all the WAL segments still available from the old S3 bucket
- reconfigure pgbackrest on the PostgreSQL node to use the new S3 bucket

pgbackrest was unable to continue restoring WAL segments due to a gap (WAL segments marked as deleted on the S3 bucket):

```
< 2020-04-10 03:55:25.335 UTC > [2062]: [1-1] user: , db: , client:  LOG:  started streaming WAL from primary at 2ED54/EE000000 on timeline 2
< 2020-04-10 03:55:25.335 UTC > [2062]: [2-1] user: , db: , client:  FATAL:  could not receive data from WAL stream: ERROR:  requested WAL segment 000000020002ED54000000EE has already been removed
```

As far as we could see all the WAL segments before the start of the first (full) backup on the new S3 bucket were expired at the end of the backup itself, compromising our recovery window.

We know that:

- the WAL segments were correctly archived by PostgreSQL:
```
2020-04-07 08:57:20.827 P00   INFO: push 46 WAL file(s) to archive: 000000020002ED54000000EE...000000020002ED550000001B
2020-04-07 08:57:21.914 P01 DETAIL: pushed WAL file '000000020002ED54000000EE' to the archive
```
 - the WAL segments were no longer available:
```
pgbackrest --stanza=db --log-level-console=detail info
stanza: db
    status: ok
    cipher: none

    db (current)
        wal archive min/max (9.6-1): 000000020002ED55000000AA/000000020002F08B0000001A
```

 - We have versioning available on the S3 bucket, hence we could see the WAL segments `000000020002ED54000000EE...000000020002ED550000001B` associated with a delete marker.

 - We can confirm that the WAL segments were expired at the end of the first full backup on the new bucket:
```
2020-04-07 13:31:04.357 P00 DETAIL: archive retention on backup 20200407-090214F, archiveId = 9.6-1, start = 000000020002ED55000000AA
2020-04-07 13:31:07.911 P00 DETAIL: remove archive: archiveId = 9.6-1, start = 000000020002ED54, stop = 000000020002ED55000000A9
```

 - We were able to continue the restore removing the delete markers, but the following backup operation expired the WAL segments in the range again.

 - We haven't found a way to mitigate this behavior, because:
   - we cannot start a backup if archiving is not already enabled
   - when we switch WAL archiving to the new bucket, all WAL segments archived before the first backup starts will eventually be expired

Any clarification on the mechanism is very welcome. If S3 versioning was not in place we might have lost the WAL segments and broken our recoverability window.

**Comment by dwsteele** [comment]
The new repo has no knowledge of the old repo so it considers any WAL before the oldest full backup to be useless from a PITR perspective. The usual strategy is to copy the old repo to the new one but with 100TB on S3 I can see why you would rather not.

You can preserve WAL before the oldest full backup by *not* setting `repo1-retention-full` or `repo1-retention-archive`. The downside is that you'll get a warning that `repo1-retention-full` is not set, but that should preserve your older WAL at least.

It would be nice if you could do this without the warning but this particular issue has never come up. Maybe something like `repo1-retention-archive=all` would be beneficial, but it's not something we are actively considering.

**Comment by acheccuc-tw** [comment]
Thanks a lot for dealing with my question and sorry for the late reply, I have been very busy last week.
I was wondering if, to cover cases like our, it wouldn't be better to expire all the WAL segments preceding the first full backup together with the backup itself. We have been lucky to have S3. versioning in place, otherwise we would have had a gap in the WAL timeline.
Alternatively, expiring by time, rather than backup copies could be helpful.

**Comment by dwsteele** [comment]
> I was wondering if, to cover cases like our, it wouldn't be better to expire all the WAL segments preceding the first full backup together with the backup itself.

We have implemented something this this at c5241e5007c4. It made sense in the context of some other expire work we are doing and it seems it would have solved your problem as well.
